### ISSUE TITLE
Converting a make target to a poe task

### ISSUE THREAD

**Issue by amotl** [nat-n/poethepoet#18]
Dear Nat,

within [wireviz-web](https://github.com/daq-tools/wireviz-web), we want convert a `make release` target [1] to a Poe the Poet task. Currently, this target can be invoked by, for example, either `make release` or `make release bump=patch`.

Therefore, we need some features which have not been obvious to us from reading the documentation. Let me share some insights about what our needs are, maybe those features are already implemented.

- We want to make this task obtain an **optional** extra argument `bump` from the command line. If it is omitted, it should use a specific default value (here: `minor`).
- We want to set a variable at runtime (here: `version`), which obtains the current package version by invoking `poetry version --short`. This variable should be available within subsequent task steps.

We started implementing this as a "shell" task [2] but are currently blocked by

- https://github.com/uiri/toml/issues/348 - due to `black` and `isort` not able to parse the given section [2] and
- #17 - the possibility to make the "shell" task obtain any extra arguments at all.

Maybe this use case sparks your interest. We will be happy if you or someone from the community will be able to do something about it.

With kind regards,
Andreas.

[1] https://github.com/daq-tools/wireviz-web/blob/f9be8b7b/Makefile#L4-L11
[2] https://github.com/daq-tools/wireviz-web/blob/f9be8b7b/pyproject.toml#L84-L92

**Comment by nat-n** [comment]
Hi @amotl,

Thanks for the feedback. It sounds like what you want is something like #6 which is high up on the list of features I'd like to add when I next get a solid block of time to work on it. So maybe you'd be able to do something along the lines of:

```toml
release.args = { bump = "minor" }
release.shell = """
poetry version {bump}
VERSION=$(shell poetry version --short))
git commit pyproject.toml -m "Bump version to $(VERSION)"
git tag $(VERSION)
git push && git push --tags
poetry build && poetry publish
"""
```
Which could then be invoked like `poe release --bump major` 

Would that be a good solution for you?

I'm afraid I'm not sure exactly when I'll have this in a release, so in the mean time possible workarounds might be to:

a) pass the *bump* value as an env var like `BUMP=major poe release` which should then be available from shell task –although maybe this defies the point of having a nice CLI

b) define the task logic in a python function which can be given as a *script* task and access given cli args from `sys.argv`. This is super powerful of course but requires another python module so maybe undermines the self containment of pyproject.toml.

---

As for the toml syntax issue, *sigh* it seems this a is fundamental problem with many tools sharing the same toml file but not the same level of toml support. Fortunately toml syntax is flexible so the following two examples are equivalent and might be better supported:

```toml
[tool.poe.tasks.release]
script = """
your multi-line script lines here
"""
help = "task description here"
```

```toml
[tool.poe.tasks]
# this one might also not work as expected with the toml library
release.script = """
your multi-line script lines here
"""
release.help = "task description here"
```

**Comment by amotl** [comment]
Hi Nat,

thanks for your quick answer and for suggesting so many valuable details.

> EDIT: It just occurred to me that a more quick win approach in terms of new functionality might be to simply support positional arguments on shell tasks which are then available a $1, $2, $@ etc within the shell task content. I guess this might also suffice for your use case?

Yes, I believe that should be sufficient and a very nice feature for `poe` in order to better support shell tasks with parameters. shell has only positional args anyway and might parse $@ how ever it is convenient, right?

> Suggestion b): Define the task logic in a python function.

For our use case, I followed that approach. Actually, both shell scripts **and** Makefile targets feel crufty to some extent when adding more and more features. These days, you never know how one might want to evolve even a small automation script just about releasing a package (notify Slack or others using [apprise](https://github.com/caronc/apprise), anyone? ;]), right?

The outcome is [1] and [2] and it works very well. I always played with the idea of rewriting that in Python but your suggestion made the final click - so, thanks! When `release.py` can be put into a separate little package on PyPI, it might even be able to be made reusable for others.

Did you actually already have the idea about providing a `poe-contrib`-like space/repository for things like that? If you think there would be a fit, please let me know.

With kind regards,
Andreas.

[1] https://github.com/daq-tools/wireviz-web/blob/0.1.0/pyproject.toml#L99-L100
[2] https://github.com/daq-tools/wireviz-web/blob/0.1.0/release.py

**Comment by nat-n** [comment]
@amotl Glad to hear it worked out.

I'd thought a little about having a poe-contrib like repo for custom task or executor types, maybe it could work for task content as well. Or maybe that would work better as an awesome list.

I'll close this issue, since it sounds like it's resolved for you except what is covered by #17.
